src/mbgl/text/glyph_set.cpp

```diff
@@ -88,81 +99,122 @@ void justifyLine(std::vector<PositionedGlyph> &positionedGlyphs, const std::map<
-            shape.y += lineFeedOffset * line;
-
-            if (shape.x > maxWidth && lastSafeBreak > 0) {
-
-                uint32_t lineLength = positionedGlyphs[lastSafeBreak + 1].x;
-                maxLineLength = util::max(lineLength, maxLineLength);
-
-                for (uint32_t k = lastSafeBreak + 1; k <= i; k++) {
-                    positionedGlyphs[k].y += lineFeedOffset;
-                    positionedGlyphs[k].x -= lineLength;
-                }
-
-                if (justify) {
-                    // Collapse invisible characters.
-                    uint32_t breakGlyph = positionedGlyphs[lastSafeBreak].glyph;
-                    uint32_t lineEnd = lastSafeBreak;
-                    if (util::i18n::isVisible(breakGlyph)) {
-                        lineEnd--;
-                    }
-
-                    justifyLine(positionedGlyphs, sdfs, lineStartIndex, lineEnd, justify);
-                }
-
-                lineStartIndex = lastSafeBreak + 1;
-                lastSafeBreak = 0;
-                lengthBeforeCurrentLine += lineLength;
-                line++;
-            }
-
-            // Ideographic characters, spaces, and word-breaking punctuation that often appear without surrounding spaces.
-            if (useBalancedIdeographicBreaking
-                || util::i18n::allowsWordBreaking(shape.glyph)
-                || util::i18n::allowsIdeographicBreaking(shape.glyph)) {
-                lastSafeBreak = i;
-            }
+        if (currentX > maxWidth && lastSafeBreak > 0) {
+            lineBreakPoints.insert(lastSafeBreak);
+            currentX -= lastSafeBreakX;
+            lastSafeBreakX = 0;
         }
+
+        currentX += glyph.metrics.advance + spacing;
     }
 
-    const PositionedGlyph& lastPositionedGlyph = positionedGlyphs.back();
-    const auto lastGlyphIt = sdfs.find(lastPositionedGlyph.glyph);
-    assert(lastGlyphIt != sdfs.end());
-    const uint32_t lastLineLength = lastPositionedGlyph.x + lastGlyphIt->second.metrics.advance;
-    maxLineLength = std::max(maxLineLength, lastLineLength);
+    return lineBreakPoints;
+}
+
+void GlyphSet::shapeLines(Shaping& shaping,
+                          const std::vector<std::u16string>& lines,
+                          const float spacing,
+                          const float lineHeight,
+                          const float horizontalAlign,
+                          const float verticalAlign,
+                          const float justify,
+                          const Point<float>& translate) const {
+
+    // the y offset *should* be part of the font metadata
+    const int32_t yOffset = -17;
+
+    float x = 0;
+    float y = yOffset;
+
+    float maxLineLength = 0;
 
-    const uint32_t height = (line + 1) * lineHeight;
+    for (std::u16string line : lines) {
+        // Collapse whitespace so it doesn't throw off justification
+        boost::algorithm::trim_if(line, boost::algorithm::is_any_of(u" \t\n\v\f\r"));
```

Comment:
They won't be trimmed, so they may affect layout slightly. Whether they allow word breaking is controlled by `i18n::allowsWordBreaking`. It looks like we don't currently break on thin/em/en spaces, either.